### DISCUSSION TITLE
Python dependency constraints

### DISCUSSION THREAD

**Discussion by amin-nejad** [renovatebot/renovate#19708] [Request Help] [answered]
Hi, I'm using the `pip_requirements` manager in my `renovate.json` to point to my `requirements.txt` file. However I have some constraints that I want to apply to the upgrades in this file e.g. `numpy<1.24`. Is there any way that I can specify this in my `renovate.json`? I can't figure out how to do this and renovate is trying to update `numpy` past this version. I have tried using the `constraints` [key](https://docs.renovatebot.com/configuration-options/#constraints) but it doesn't appear that this is supported here

**Comment by viceice** [comment, marked as answer]
use a package rule with allowed version

**Reply by amin-nejad** [reply]
Thanks very much! Is there anyway to point to a file of these constraints by any chance?

**Reply by amin-nejad** [reply]
In case of use to others in future, I added this item within `packageRules`:

```json
    {
      "matchPackageNames": ["numpy"],
      "allowedVersions": "<1.24"
    }
```
